specification/under_development/profiles/Base_Profile.md

```diff
@@ -0,0 +1,262 @@
+  the program to store measurement outcomes
+
+These attributes will show up as an [attribute
+group](https://releases.llvm.org/13.0.1/docs/LangRef.html#attrgrp) in the IR.
+Attribute groups are numbered such that they can be easily referenced by
+multiple function definitions or global variables. Arbitrary custom attributes
+may be optionally attached to any of the declared functions to convey additional
+information about that function. Consumers of Base Profile compliant programs
+should hence not rely on the numbering of the entry point attribute group, but
+instead look for function to which an attribute with the name `entry_point` is
+attached to determine which one to invoke when the program is launched.
+
+To indicate the total number of qubits required to execute the program, a custom
+attribute with the name `required_qubits` is defined and attached to the entry
+point. To indicate the number of registers/bits needed to store measurement
+results during program execution, a custom attribute with the name
+`required_results` is defined and attached to the entry point. The value of both
+of these attributes is the string representation of a 64-bit integer constant.
+
+TODO: why the string value, and not an integer? TODO: no double underscore guard
+for attributes likely makes sense
+
+### Function Body
+
+The function body consists of a single block named `entry`. This implies in
+particular that no branching is permitted inside a function's body. In the
+`entry` block, any number of calls to QIS functions may be performed. To be
+compatible with the Base Profile these functions must return void. Any arguments
+to invoke them must be inlined into the call itself; they hence must be
+constants or a pointer to a [qubit or result]() The result of the program
+execution should be logged using output recording functions.
+
+The following instructions are the *only* LLVM instructions that are permitted
+within a Base Profile compliant program:
+
+| LLVM Instruction         | Context and Purpose                                                                                              | Rules for Usage                                                                                       |
+| :----------------------- | :--------------------------------------------------------------------------------------------------------------- | :---------------------------------------------------------------------------------------------------- |
+| `call`                   | Used within a function block to invoke any one of the declared QIS functions and the output recording functions. | May optionally be preceded by a [`tail` marker](https://llvm.org/docs/LangRef.html#call-instruction). |
+| `ret`                    | Used to return the exit code of the program.                                                                     | Must occur (only) as the final instruction of the `entry` block.                                      |
```

Comment:
Indeed, as defined I don't believe there is a way to return an exit code that is not a compile time constant. I don't believe the Base Profile allows for any kind of dynamic failure. Maybe I am missing what you had in mind with the last sentence/question? 

I should also clarify that the exit code is a way to avoid needing support for early termination of the execution for cases where one would usually expect a runtime exception (e.g. division by zero) due to user/program error; i.e. it has a purpose for more advanced profiles, but not really for the Base Profile. The exist code is part of the program output, and specifically, each shot would potentially have a different exit code (histograms can e.g. filter based on exist code). If the job as a whole fails for some reason, or execution fails even though there is nothing that should fail in the program, then there should be a separate handling as part of the job APIs for that. In principle it may be conceivable that there is some handling for the execution "randomly failing" on the QPU on a per-shot basis, and in principle that could be reflected in the produced output. However, given that such a failure in that case would have nothing to do with the program itself, I don't see why that possibility would be reflected in the IR.